bokeh/plotting/helpers.py

```diff
@@ -278,34 +278,75 @@ def _pop_renderer_args(kwargs):
     return result
 
 
-def _pop_colors_and_alpha(glyphclass, kwargs, prefix="", default_alpha=1.0):
+def _pop_visuals(glyphclass, props, prefix="", defaults={}, trait_defaults={}):
     """
-    Given a kwargs dict, a prefix, and a default value, looks for different
-    color and alpha fields of the given prefix, and fills in the default value
-    if it doesn't exist.
+    Applies basic cascading logic to deduce properties for a glyph.
+
+    Args:
+        glyphclass :
+            the type of glyph being handled
+
+        props (dict) :
+            Maps properties and prefixed properties to their values.
+            Keys in `props` matching `glyphclass` properties with added `prefix`
+              will get popped, other keys will be ignored.
+            Keys take the form '[{prefix}][{feature}_]{trait}'. Only {feature}
+              must not containt underscores.
+            Keys of the form '{prefix}{trait}' work as lower precedence aliases
+              for all {features}.
+            Ex: {'fill_color': 'blue', 'selection_line_width': 0.5}
+
+        prefix (str) :
+            Prefix used when accessing `props`. Ex: 'selection_'
+
+        defaults (dict) :
+            Property fallback, in case prefixed property not in `props`.
+            Ex. 'line_width' here may be used for 'selection_line_width'.
+
+        trait_defaults (dict) :
+            Fallback based on '{trait}', in case property not in `defaults`.
+            Ex. 'width' here may be used for 'selection_line_width'.
+
+    Returns:
+        result (dict) :
+            Resulting properties for the instance (no prefixes).
+
+    Notes:
+        Feature trait 'text_color', as well as traits 'color' and 'alpha', have
+        ultimate defaults in case those can't be deduced.
     """
+    def split_feature_trait(ft):
+        """Feature is up to first '_'. Ex. 'line_color' => ['line', 'color']"""
+        ft = ft.split('_', 1)
+        return ft if len(ft)==2 else ft+[None]
+
+    def is_visual(ft):
+        """Whether a feature trait name is visual"""
+        feature, trait = split_feature_trait(ft)
+        return feature in ('line', 'fill', 'text') and trait is not None
+
     result = dict()
+    defaults = dict( [('text_color', 'black')] + list(defaults.items()) )
+    trait_defaults = dict( [('color', get_default_color()), ('alpha', 1.0)]
+                          + list(trait_defaults.items()) )
```

Comment:
I'm not sure I understand your comment. The current code is the exact equivalent of:
```
defaults = defaults.copy()
defaults.setdefault('text_color', 'black')
trait_defaults = trait_defaults.copy()
trait_defaults.setdefault('color', get_default_color())
trait_defaults.setdefault('alpha', 1.0)
```
My feeling was that, since we're required to create new dictionaries because we don't want to modify those passed as defaults, we might as well state their contents in the constructor.

I'll be happy to use the `setdefault` version if you still prefer it, let me know!

Reply:
Ah it's not obvious (or known before) how the `dict` constructor responds in the presence of repeated keys in tuples. I'd personally prefer the `setdefault` version as clearer (to me)